### PR TITLE
feat: implement update cart quantity and delete cart item

### PR DESCRIPTION
**Pull request by kibo-susantasamanta**

### CODE REVIEW

**Comment by sss-bct** [comment]
Development is done, after review will add test cases.

File: hooks/mutations/useCheckoutMutation/useCheckoutMutation.tsx [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+  const response = await client.request({
+    document: getOrCreateCheckoutFromCartMutation,
+    variables: { cartId },
+  })
+
+  return response?.checkout
+}
+
+export const useCheckoutMutation = () => {
```
Comment (kibo-kevinwatts) [inline hooks/mutations/useCheckoutMutation/useCheckoutMutation.tsx:17]: this mutation doesn't really belong to "checkout" or "cart"
rename to "useCreateFromCartMutation"
like 
```
export const useCreateFromCartMutation = () => { 
  return {
    createFromCart: useMutation(getOrCreateCheckout)
  }
}
```

thoughts @kibo-sushant ?

File: components/common/QuantitySelector/QuantitySelector.tsx [hunk 2]
```diff
@@ -27,27 +29,49 @@ const styles = {
...
-      style: {
-        paddingTop: 5,
-        paddingBottom: 5,
-        textAlign: 'center',
-      },
-    }}
-    sx={{ width: '46px', height: '37px' }}
-  />
-)
+const QuantityTextField = ({ quantity, handleCustomQuantity }: QuantityInputProps) => {
```
Comment (kibo-kevinwatts) [inline components/common/QuantitySelector/QuantitySelector.tsx:32]: how does 'handleCustomQuantity' differ from 'setItemQuantity'?

**Review by kibo-kevinwatts — COMMENTED** [review]
good job @sss-bct!

File: hooks/mutations/useCheckoutMutation/useCheckoutMutation.tsx (hunk 1 above)
Comment (sss-bct) [reply hooks/mutations/useCheckoutMutation/useCheckoutMutation.tsx]: Ok. Resolved.

File: components/common/QuantitySelector/QuantitySelector.tsx (hunk 2 above)
Comment (sss-bct) [reply components/common/QuantitySelector/QuantitySelector.tsx:32]: handleCustomQuantity is a callback through props to make API call happen from parent where as setItemQuantity is used to update an internal state of quantity to make QuantityTextField a controlled one.

File: hooks/mutations/useCartMutation/useCartMutation.ts [hunk 3]
```diff
@@ -33,9 +45,59 @@ const addToCart = async (props: AddToCartInputParams) => {
...
+    const response = await client.request({
+      document: updateCartItemQuantityMutation,
+      variables,
+    })
+
+    return response?.updateCartItemQuantity
+  } catch (err) {
+    console.error(err)
+  }
+}
```
Comment (kibo-sushant) [inline hooks/mutations/useCartMutation/useCartMutation.ts:66]: We already have global error handler, do we need to use Try Catch block here?

File: hooks/mutations/useCartMutation/useCartMutation.ts [hunk 4]
```diff
@@ -33,9 +45,59 @@ const addToCart = async (props: AddToCartInputParams) => {
...
+      onSuccess: () => {
+        queryClient.invalidateQueries(cartKeys.all)
+      },
+    }),
+
+    removeCartItem: useMutation(removeCartItem, {
+      onSuccess: () => {
+        queryClient.invalidateQueries(cartKeys.all)
+      },
+    }),
```
Comment (kibo-sushant) [inline hooks/mutations/useCartMutation/useCartMutation.ts:101]: I think we should follow the single responsibility principle here.
One custom hook should have one responsiblity, need to break hooks separate files

File: components/common/QuantitySelector/QuantitySelector.tsx [hunk 5]
```diff
@@ -27,27 +29,45 @@ const styles = {
...
+  const [itemQuantity, setItemQuantity] = useState<number | string>(quantity)
```
[the remark this answers is not in the document]
Comment (sss-bct) [reply components/common/QuantitySelector/QuantitySelector.tsx:33]: value coming from event.target is a string, for that it's required.

File: hooks/mutations/useCartMutation/useCartMutation.ts (hunk 3 above)
Comment (sss-bct) [reply hooks/mutations/useCartMutation/useCartMutation.ts:62]: Removed.

File: hooks/mutations/useCartMutation/useCartMutation.ts (hunk 4 above)
Comment (sss-bct) [reply hooks/mutations/useCartMutation/useCartMutation.ts:129]: But it is working fine and each mutation is taking single responsibility here. As we decided I followed this approach, if you think it's mandatory I will update.

Comment (kibo-sushant) [reply hooks/mutations/useCartMutation/useCartMutation.ts:129]: Small code is always easy to understand and more maintainable. One file should have only one hook and one hook should have only one responsibility.

Comment (sss-bct) [reply hooks/mutations/useCartMutation/useCartMutation.ts:129]: I will break it up and all the remaining items will be coved as well.

**Review by kibo-sushant — APPROVED** [review]

File: components/common/QuantitySelector/QuantitySelector.tsx [hunk 6]
```diff
@@ -27,27 +29,50 @@ const styles = {
...
+
+  const handleChangeQuantityOnBlur = () => {
+    if (itemQuantity !== '' && itemQuantity !== quantity) {
+      handleCustomQuantity(Number(itemQuantity))
+    } else setItemQuantity(quantity)
+  }
+
+  useEffect(() => {
+    setItemQuantity(quantity)
+  }, [quantity])
```
Comment (kibo-kevinwatts) [inline components/common/QuantitySelector/QuantitySelector.tsx:50]: I don't think we need this, quantity is coming in as prop and assigned to state

File: components/common/QuantitySelector/QuantitySelector.tsx [hunk 7]
```diff
@@ -27,27 +29,45 @@ const styles = {
...
+    const newQuantity = Number(e.target.value)
+
+    if (!Number.isNaN(newQuantity) && newQuantity > 0 && newQuantity !== quantity) {
+      handleCustomQuantity(Number(itemQuantity))
+    } else setItemQuantity(quantity)
+  }
+
+  useEffect(() => {
+    setItemQuantity(quantity)
+  }, [quantity])
```
[the remark this answers is not in the document]
Comment (kibo-kevinwatts) [reply components/common/QuantitySelector/QuantitySelector.tsx:50]: Agree with sushant, i don't think the validation for the quantity is correct here.

File: __test__/e2e/integration/page-template/cart/CartTemplate.spec.js [hunk 8]
```diff
@@ -49,4 +40,30 @@ describe('[components] CartTemplate integration', () => {
...
+    const { user } = setup(props)
+
+    const cartItem = screen.getAllByRole('group')
+    expect(cartItem).toHaveLength(2)
+
+    const deleteButton = screen.getAllByRole('button', { name: 'item-delete' })
+    await user.click(deleteButton[0])
+    const deletedCartItem = screen.queryByText('Pink Backpack')
+    await waitFor(() => expect(deletedCartItem).not.toBeInTheDocument())
+  })
```
Comment (kibo-kevinwatts) [inline __test__/e2e/integration/page-template/cart/CartTemplate.spec.js:68]: test case for decrease quantity?

**Review by kibo-kevinwatts — CHANGES_REQUESTED** [review]

File: hooks/mutations/useCartMutation/useCartMutation.ts [hunk 9]
```diff
@@ -1,10 +1,15 @@
...
+import {
+  addToCartMutation,
+  updateCartItemQuantityMutation,
+  deleteCartItemMutation,
+} from '@/lib/gql/mutations'
 import { buildAddToCartInput } from '@/lib/helpers/buildAddToCartInput'
+import { cartKeys } from '@/lib/react-query/queryKeys'
 
-import { ProductOption } from '@/lib/gql/types'
+import { CartItem, ProductOption } from '@/lib/gql/types'
```
[the remark this answers is not in the document]
Comment (sss-bct) [reply hooks/mutations/useCartMutation/useCartMutation.ts]: Done

File: components/common/QuantitySelector/QuantitySelector.tsx (hunk 6 above)
Comment (sss-bct) [reply components/common/QuantitySelector/QuantitySelector.tsx]: Actually it was not  rendering automatically after API response but found a solution with key. So Remove. Thank you.

File: __test__/e2e/integration/page-template/cart/CartTemplate.spec.js (hunk 8 above)
Comment (sss-bct) [reply __test__/e2e/integration/page-template/cart/CartTemplate.spec.js:80]: Done.

**Review by kibo-kevinwatts — APPROVED** [review]

**Pull request merged**